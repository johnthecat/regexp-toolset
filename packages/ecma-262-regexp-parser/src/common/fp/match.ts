export type Matched<T> = {
  match: true;
  value: T;
};

export type Unmatched = {
  match: false;
};

export type UnmatchedError = {
  match: false;
  error: Error;
};

export interface Match<M> {
  map<R1>(fn: (value: M) => R1): Match<R1>;

  orElse<R>(fn: () => Match<R>): Match<M | R>;

  orError(fn: () => Error): Match<M>;

  match<R1>(fn: (value: M) => Match<R1>): Match<R1>;

  unmatch<R1>(fn: () => Match<R1>): Match<R1>;

  isMatched(): Match<boolean>;

  unwrap(): Matched<M> | Unmatched | UnmatchedError;

  filter<R extends M>(fn: (value: M) => value is R): Match<R>;

  filter(fn: (value: M) => boolean): Match<M>;

  filterOrThrow<R extends M>(fn: (value: M) => value is R, err: () => Error): Match<R>;

  filterOrThrow(fn: (value: M) => boolean, err: (value: M) => Error): Match<M>;
}

type InferMatchedValue<T> = T extends Match<infer U> ? U : never;
type Result<M> = Matched<M> | Unmatched | UnmatchedError;

const isMatched = <M>(x: Result<M>): x is Matched<M> => x.match;
const isUnmatched = <M>(x: Result<M>): x is Unmatched => !isMatched(x) && !('error' in x);

const caught = <T>(fn: () => Match<T>): Match<T> => {
  try {
    return fn();
  } catch (e) {
    return err(e as Error) as unknown as Match<T>;
  }
};

const matchImplementation = <M>(x: Result<M>): Match<M> => {
  const self: Match<M> = {
    match: fn => (isMatched(x) ? caught(() => fn(x.value)) : (self as any)),
    unmatch: fn => (isUnmatched(x) ? fn() : (self as Match<never>)),
    map: fn => (isMatched(x) ? caught(() => ok(fn(x.value))) : (self as Match<never>)),
    filter: (fn: (x: M) => boolean) =>
      isMatched(x) ? caught(() => (fn(x.value) ? (self as Match<never>) : none())) : (self as Match<never>),
    filterOrThrow: (fn: (x: M) => boolean, error: (x: M) => Error) =>
      isMatched(x)
        ? caught(() => (fn(x.value) ? (self as Match<never>) : err(error(x.value))))
        : (self as Match<never>),
    isMatched: () => ok(x.match),
    orElse: fn => (isUnmatched(x) ? fn() : self),
    orError: fn => (isUnmatched(x) ? err(fn()) : self),
    unwrap: () => x,
  };

  return self;
};

export const ok = <M>(value: M): Match<M> => {
  return matchImplementation({
    match: true,
    value,
  });
};

export const none = (): Match<never> => {
  return matchImplementation({ match: false });
};

export const err = (error: Error): Match<never> => {
  return matchImplementation({ match: false, error });
};

export const all = <T extends Match<unknown>[]>(
  matches: [...T],
): Match<{ [K in keyof T]: InferMatchedValue<T[K]> }> => {
  const result: unknown[] = [];
  for (const matcher of matches) {
    const v = matcher.unwrap();
    if (!v.match) {
      return matcher as never;
    }
    result.push(v.value);
  }

  return ok(result) as Match<{ [K in keyof T]: InferMatchedValue<T[K]> }>;
};

export const first = <R1>(...matchers: (() => Match<R1>)[]): Match<R1> => {
  let lastResult: Match<R1> = none();
  for (const matcher of matchers) {
    lastResult = lastResult.unmatch(matcher);
  }
  return lastResult.orError(() => new Error(`Can't match any.`));
};

export const nonNullable = <T>(value: T): Match<NonNullable<T>> =>
  value === null || value === void 0 ? none() : ok(value);
