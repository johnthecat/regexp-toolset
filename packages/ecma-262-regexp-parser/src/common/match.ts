export type Matched<T> = {
  match: true;
  value: T;
};

export type Unmatched<T> = {
  match: false;
  value: T;
};

export type UnmatchedError = {
  match: false;
  error: Error;
};

export type Result<M, U = M> = Matched<M> | Unmatched<U> | UnmatchedError;

export type Monad<M, U = M> = {
  matched<R1, R2>(fn: (value: M) => Monad<R1, R2>): Monad<R1, R2>;
  unmatched<R1, R2>(fn: (value: U) => Monad<R1, R2>): Monad<R1, R2>;
  map<R1>(fn: (value: M) => R1): Monad<R1, U>;
  flatMap<R1, R2>(fn: (result: Result<M, U>) => Monad<R1, R2>): Monad<R1, R2>;
  error<R1, R2>(fn: (error: Error) => Monad<R1, R2>): Monad<R1, R2>;
  unwrap(): Result<M, U>;
};

type InferMonadMatchedValue<T> = T extends Monad<infer U, unknown> ? U : never;

export const matched = <M>(value: M): Monad<M, never> => {
  const self: Monad<M, never> = {
    matched: fn => fn(value),
    unmatched: () => self as Monad<never>,
    error: () => self as Monad<never>,
    map: fn => matched(fn(value)),
    flatMap: fn => fn(self.unwrap()),
    unwrap: () => ({
      match: true,
      value,
    }),
  };

  return self;
};

export const unmatched = <U>(value: U): Monad<never, U> => {
  const self: Monad<never, U> = {
    matched: () => self as Monad<never>,
    unmatched: fn => fn(value),
    error: () => self as Monad<never>,
    map: () => self,
    flatMap: fn => fn(self.unwrap()),
    unwrap: () => ({
      match: false,
      value,
    }),
  };

  return self;
};

export const errored = (error: Error): Monad<never> => {
  const self: Monad<never> = {
    matched: () => self,
    unmatched: () => self,
    error: fn => fn(error),
    map: () => self,
    flatMap: fn => fn(self.unwrap()),
    unwrap: () => ({
      match: false,
      error,
    }),
  };

  return self;
};

export const matchSeq = <T extends Monad<unknown>[]>(
  matches: [...T],
): Monad<{ [K in keyof T]: InferMonadMatchedValue<T[K]> }> => {
  let result = matched<unknown[]>([]);

  for (const monad of matches) {
    result = monad.matched(x => result.map(list => list.concat(x)));
  }

  return result as Monad<{ [K in keyof T]: InferMonadMatchedValue<T[K]> }>;
};

export const matchFirst = <R1>(
  matchers: (() => Monad<R1, unknown>)[],
  defaultReturn?: Monad<R1, unknown>,
): Monad<R1, unknown> => {
  let lastResult: Monad<R1, unknown> = unmatched(null);
  for (const matcher of matchers) {
    lastResult = lastResult.unmatched(() => matcher());
  }
  return lastResult.unmatched(() => defaultReturn ?? errored(new Error(`Can't match any.`)));
};
