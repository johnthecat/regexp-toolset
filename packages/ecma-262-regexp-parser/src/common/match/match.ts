import type { Match, Matched, Unmatched, UnmatchedError } from './types.js';

type InferMatchedValue<T> = T extends Match<infer U> ? U : never;
type Result<M> = Matched<M> | Unmatched | UnmatchedError;

const isMatched = <M>(x: Result<M>): x is Matched<M> => x.match;
const isError = <M>(x: Result<M>): x is UnmatchedError => !isMatched(x) && 'error' in x;
const isUnmatched = <M>(x: Result<M>): x is Unmatched => !isMatched(x) && !('error' in x);

const matchImplementation = <M>(x: Result<M>): Match<M> => {
  const self: Match<M> = {
    matched: fn => self.flatMap(fn),
    unmatched: fn => (isUnmatched(x) ? fn() : (self as Match<never>)),
    error: fn => (isError(x) ? fn(x.error) : (self as Match<never>)),
    map: fn => (isMatched(x) ? ok(fn(x.value)) : (self as Match<never>)),
    flatMap: fn => (isMatched(x) ? fn(x.value) : (self as any)),
    filter: (fn: (x: M) => boolean) =>
      isMatched(x) ? (fn(x.value) ? (self as Match<never>) : none()) : (self as Match<never>),
    isMatched: () => ok(x.match),
    orElse: fn => (isUnmatched(x) ? fn() : self),
    orError: fn => (isUnmatched(x) ? err(fn()) : self),
    unwrap: () => x,
  };

  return self;
};

export const ok = <M>(value: M): Match<M> => {
  return matchImplementation({
    match: true,
    value,
  });
};

export const none = (): Match<never> => {
  return matchImplementation({ match: false });
};

export const err = (error: Error): Match<never> => {
  return matchImplementation({ match: false, error });
};

export const all = <T extends Match<unknown>[]>(
  ...matches: [...T]
): Match<{ [K in keyof T]: InferMatchedValue<T[K]> }> => {
  const result: unknown[] = [];
  for (const matcher of matches) {
    const v = matcher.unwrap();
    if (!v.match) {
      return matcher as never;
    }
    result.push(v.value);
  }

  return ok(result) as Match<{ [K in keyof T]: InferMatchedValue<T[K]> }>;
};

export const first = <R1>(...matchers: (() => Match<R1>)[]): Match<R1> => {
  let lastResult: Match<R1> = none();
  for (const matcher of matchers) {
    lastResult = lastResult.unmatched(matcher);
  }
  return lastResult.orError(() => new Error(`Can't match any.`));
};

export const nonNullable = <T>(value: T): Match<NonNullable<T>> =>
  value === null || value === void 0 ? none() : ok(value);

export { Match };
