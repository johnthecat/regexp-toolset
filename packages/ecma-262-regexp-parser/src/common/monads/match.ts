import type { Match, Matched, Unmatched, UnmatchedError } from './types.js';

type InferMatchedValue<T> = T extends Match<infer U> ? U : never;
type Result<M> = Matched<M> | Unmatched | UnmatchedError;

const isError = <M>(x: Result<M>): x is UnmatchedError => {
  return !x.match && 'error' in x;
};
const isUnmatched = <M>(x: Result<M>): x is Unmatched => {
  return !x.match && !('error' in x);
};
const isMatched = <M>(x: Result<M>): x is Matched<M> => {
  return x.match;
};

const matchImplementation = <M>(x: Result<M>): Match<M> => {
  const self: Match<M> = {
    matched: fn => self.flatMap(fn),
    unmatched: fn => (isUnmatched(x) ? fn() : (self as Match<never>)),
    error: fn => (isError(x) ? fn(x.error) : (self as Match<never>)),
    map: fn => (isMatched(x) ? matched(fn(x.value)) : (self as Match<never>)),
    flatMap: fn => (isMatched(x) ? fn(x.value) : (self as any)),
    filter: fn => (isMatched(x) ? (fn(x.value) ? (self as Match<never>) : unmatched()) : (self as Match<never>)),
    isMatched: () => matched(x.match),
    orElse: fn => (isUnmatched(x) ? fn() : self),
    orError: fn => (isUnmatched(x) ? errored(fn()) : self),
    unwrap: () => x,
  };

  return self;
};

export const matched = <M>(value: M): Match<M> => {
  return matchImplementation({
    match: true,
    value,
  });
};

export const unmatched = (): Match<never> => {
  return matchImplementation({ match: false });
};

export const errored = (error: Error): Match<never> => {
  return matchImplementation({ match: false, error });
};

export const all = <T extends Match<unknown>[]>(
  ...matches: [...T]
): Match<{ [K in keyof T]: InferMatchedValue<T[K]> }> => {
  const result: unknown[] = [];
  for (const matcher of matches) {
    const v = matcher.unwrap();
    if (!v.match) {
      return matcher as never;
    }
    result.push(v.value);
  }

  return matched(result) as Match<{ [K in keyof T]: InferMatchedValue<T[K]> }>;
};

export const first = <R1>(...matchers: (() => Match<R1>)[]): Match<R1> => {
  let lastResult: Match<R1> = unmatched();
  for (const matcher of matchers) {
    lastResult = lastResult.unmatched(matcher);
  }
  return lastResult.orError(() => new Error(`Can't match any.`));
};

export const nonNullable = <T>(value: T): Match<NonNullable<T>> =>
  value === null || value === void 0 ? unmatched() : matched(value);

export { Match };
