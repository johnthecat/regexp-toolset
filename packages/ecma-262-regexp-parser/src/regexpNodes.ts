export type NodePosition = { start: number; end: number };

export enum SyntaxKind {
  Regexp,
  LineStart,
  LineEnd,
  Disjunction,
  CharRange,
  Alternative,
  Char,
  ASCIIControlChar,
  ControlEscapeChar,
  NullChar,
  Backspace,
  Subpattern,
  AnyChar,
  AnyDigit,
  NonDigit,
  AnyWhitespace,
  NonWhitespace,
  AnyWord,
  NonWord,
  WordBoundary,
  NonWordBoundary,
  ZeroLength,
  CharClass,
  Group,
  GroupName,
  BackReference,
  Quantifier,
  Repetition,
}

export enum QuantifierType {
  NoneOrSingle,
  NoneOrMany,
  SingleOrMany,
  Range,
}

export enum ControlEscapeCharType {
  NewLine,
  CarriageReturn,
  Tab,
  VerticalWhitespace,
  FormFeedChar,
}

export type Node<
  Kind extends SyntaxKind = SyntaxKind,
  Value extends Record<string, unknown> = Record<string, never>,
> = Value &
  NodePosition & {
    kind: Kind;
    __value?: Value;
  };

export type InferNodeKind<T> = T extends Node<infer U, Record<string, unknown>> ? U : never;
export type InferNodeValue<T> = T extends Node<SyntaxKind, infer U extends Record<string, unknown>> ? U : never;
export type InferNodeActualValue<T> = InferNodeValue<T> extends Record<string, never> ? void : InferNodeValue<T>;

export type RegexpNode = Node<SyntaxKind.Regexp, { body: AnyRegexpNode; flags: string }>;
export type LineStartNode = Node<SyntaxKind.LineStart>;
export type LineEndNode = Node<SyntaxKind.LineEnd>;
export type DisjunctionNode = Node<SyntaxKind.Disjunction, { left: AnyRegexpNode; right: AnyRegexpNode }>;
export type CharRangeNode = Node<SyntaxKind.CharRange, { from: CharNode; to: CharNode }>;
export type AlternativeNode = Node<SyntaxKind.Alternative, { expressions: AnyRegexpNode[] }>;
export type CharNode = Node<
  SyntaxKind.Char,
  { value: string; charCode: number; type: 'simple' | 'hex' | 'unicode' | 'escaped' | 'octal' }
>;
export type ASCIIControlCharNode = Node<SyntaxKind.ASCIIControlChar, { value: string }>;
export type NullCharNode = Node<SyntaxKind.NullChar>;
export type BackspaceNode = Node<SyntaxKind.Backspace>;
export type SubpatternNode = Node<SyntaxKind.Subpattern, { groupName: string; ref: GroupNode | null }>;
export type AnyCharNode = Node<SyntaxKind.AnyChar>;
export type AnyDigitNode = Node<SyntaxKind.AnyDigit>;
export type NonDigitNode = Node<SyntaxKind.NonDigit>;
export type AnyWhitespaceNode = Node<SyntaxKind.AnyWhitespace>;
export type NonWhitespaceNode = Node<SyntaxKind.NonWhitespace>;
export type AnyWordNode = Node<SyntaxKind.AnyWord>;
export type NonWordNode = Node<SyntaxKind.NonWord>;
export type WordBoundaryNode = Node<SyntaxKind.WordBoundary>;
export type NonWordBoundaryNode = Node<SyntaxKind.NonWordBoundary>;
export type ControlEscapeCharNode = Node<SyntaxKind.ControlEscapeChar, { type: ControlEscapeCharType }>;
export type ZeroLengthNode = Node<SyntaxKind.ZeroLength>;
export type CharClassNode = Node<SyntaxKind.CharClass, { negative: boolean; expressions: AnyRegexpNode[] }>;
export type GroupNameNode = Node<SyntaxKind.GroupName, { name: string }>;
export type GroupNode = Node<
  SyntaxKind.Group,
  {
    body: AnyRegexpNode;
    specifier: GroupNameNode | null;
    type:
      | 'capturing'
      | 'nonCapturing'
      | 'positiveLookahead'
      | 'negativeLookahead'
      | 'positiveLookbehind'
      | 'negativeLookbehind';
  }
>;
export type BackReferenceNode = Node<SyntaxKind.BackReference, { group: GroupNode }>;

type QuantifierNodeRangeValue = { from: number; to?: number | void };

export type QuantifierNode = Node<
  SyntaxKind.Quantifier,
  (
    | { type: QuantifierType.NoneOrMany | QuantifierType.NoneOrSingle | QuantifierType.SingleOrMany }
    | ({ type: QuantifierType.Range } & QuantifierNodeRangeValue)
  ) & {
    greedy: boolean;
  }
>;

export type RepetitionNode = Node<SyntaxKind.Repetition, { expression: AnyRegexpNode; quantifier: QuantifierNode }>;

export type AnyRegexpNode =
  | RegexpNode
  | AlternativeNode
  | QuantifierNode
  | RepetitionNode
  | LineStartNode
  | LineEndNode
  | DisjunctionNode
  | CharRangeNode
  | AnyCharNode
  | CharNode
  | ASCIIControlCharNode
  | NullCharNode
  | BackspaceNode
  | SubpatternNode
  | CharClassNode
  | GroupNameNode
  | GroupNode
  | BackReferenceNode
  | ZeroLengthNode
  | ControlEscapeCharNode
  | AnyWhitespaceNode
  | NonWhitespaceNode
  | AnyDigitNode
  | NonDigitNode
  | AnyWordNode
  | NonWordNode
  | WordBoundaryNode
  | NonWordBoundaryNode;
