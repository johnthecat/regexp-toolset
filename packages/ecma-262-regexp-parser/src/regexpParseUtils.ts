import type { Step } from './regexpTokenizer.js';
import type { NodeParser, NodeParserResult, NodeParserResultValue, ParserContext } from './regexpParseTypes.js';
import { ok } from './common/match/match.js';

export const fillExpressions = (token: Step, ctx: ParserContext, tokenParser: NodeParser): NodeParserResult => {
  let currentParserResult: NodeParserResultValue = { nodes: [], token };

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const parseResult = tokenParser(currentParserResult, ctx);
    const unwrapped = parseResult.unwrap();
    if (unwrapped.match) {
      const nextToken = unwrapped.value.token.next();
      if (!nextToken) {
        return parseResult;
      }
      currentParserResult = { nodes: unwrapped.value.nodes, token: nextToken };
    } else {
      if ('error' in unwrapped) {
        return parseResult;
      }
      return ok(currentParserResult);
    }
  }
};
