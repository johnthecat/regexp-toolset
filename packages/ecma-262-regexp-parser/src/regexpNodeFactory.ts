import type {
  Node,
  AlternativeNode,
  AnyRegexpNode,
  BackReferenceNode,
  CharClassNode,
  CharNode,
  CharRangeNode,
  ASCIIControlCharNode,
  DisjunctionNode,
  GroupNameNode,
  GroupNode,
  InferNodeActualValue,
  InferNodeKind,
  InferNodeValue,
  NodePosition,
  QuantifierNode,
  RepetitionNode,
  RegexpNode,
  ControlEscapeCharNode,
  ControlEscapeCharType,
} from './regexpNodes.js';
import { SyntaxKind } from './regexpNodes.js';
// FIXME circular dep
import { sealExpressions } from './regexpParseUtils.js';

export const createNode = <T extends Node<SyntaxKind, Record<string, unknown>>>(
  kind: InferNodeKind<T>,
  { start, end }: NodePosition,
  value: InferNodeActualValue<T>,
): T =>
  ({
    kind,
    ...value,
    start,
    end,
  } as T);

export const createSimpleNode = <T extends Node>(kind: InferNodeKind<T>, { start, end }: NodePosition): T =>
  ({
    kind,
    start,
    end,
  } as unknown as T);

export const createRegexpNode = (expressions: AnyRegexpNode[], flags: string = ''): RegexpNode => {
  const body = sealExpressions(expressions);
  return createNode<RegexpNode>(
    SyntaxKind.Regexp,
    {
      start: body.start - 1,
      end: body.end + 1 + flags.length,
    },
    {
      flags,
      body,
    },
  );
};

export const createCharNode = (value: string, position: NodePosition, type: CharNode['type']) =>
  createNode<CharNode>(SyntaxKind.Char, position, {
    type,
    value,
    charCode: value.charCodeAt(0),
  });

export const createASCIIControlCharNode = (value: string, position: NodePosition) =>
  createNode<ASCIIControlCharNode>(SyntaxKind.ASCIIControlChar, position, {
    value,
  });

export const createAlternativeNode = (expressions: AnyRegexpNode[]) =>
  createNode<AlternativeNode>(
    SyntaxKind.Alternative,
    {
      start: expressions.at(0)?.start ?? 0,
      end: expressions.at(-1)?.end ?? 0,
    },
    {
      expressions,
    },
  );

export const createCharRangeNode = (from: CharNode, to: CharNode) =>
  createNode<CharRangeNode>(
    SyntaxKind.CharRange,
    {
      start: from.start,
      end: to.end,
    },
    {
      from: from,
      to: to,
    },
  );

export const createQuantifierNode = (position: NodePosition, value: InferNodeValue<QuantifierNode>) =>
  createNode<QuantifierNode>(SyntaxKind.Quantifier, position, value);

export const createRepetitionNode = (expression: AnyRegexpNode, quantifier: QuantifierNode) =>
  createNode<RepetitionNode>(
    SyntaxKind.Repetition,
    {
      start: expression.start,
      end: quantifier.end,
    },
    {
      quantifier,
      expression,
    },
  );

export const createBackReferenceNode = (position: NodePosition, group: GroupNode) =>
  createNode<BackReferenceNode>(SyntaxKind.BackReference, position, {
    group,
  });

export const createDisjunctionNode = (left: AnyRegexpNode[], right: AnyRegexpNode[], position: NodePosition) => {
  const normalizedPosition = {
    start: left.at(0)?.start ?? position.start,
    end: right.at(-1)?.end ?? position.end,
  };

  return createNode<DisjunctionNode>(SyntaxKind.Disjunction, normalizedPosition, {
    left: sealExpressions(left, normalizedPosition, normalizedPosition),
    right: sealExpressions(right, normalizedPosition, normalizedPosition),
  });
};

export const createCharClassNode = (negative: boolean, expressions: AnyRegexpNode[], position: NodePosition) =>
  createNode<CharClassNode>(SyntaxKind.CharClass, position, {
    negative,
    expressions,
  });

export const createGroupNameNode = (name: string, position: NodePosition) =>
  createNode<GroupNameNode>(SyntaxKind.GroupName, position, { name });

export const createControlEscapeNode = (type: ControlEscapeCharType, position: NodePosition) =>
  createNode<ControlEscapeCharNode>(SyntaxKind.ControlEscapeChar, position, { type });

export const createGroupNode = (
  type: GroupNode['type'],
  specifier: GroupNode['specifier'],
  expressions: AnyRegexpNode[],
  position: NodePosition,
) => {
  const bodyPosition = {
    start: expressions.at(0)?.start ?? (specifier ? specifier.end + 1 : position.start + 1),
    end: expressions.at(-1)?.end ?? (specifier ? specifier.end : position.start),
  };

  return createNode<GroupNode>(SyntaxKind.Group, position, {
    type,
    specifier,
    body: sealExpressions(expressions, bodyPosition, bodyPosition),
  });
};
