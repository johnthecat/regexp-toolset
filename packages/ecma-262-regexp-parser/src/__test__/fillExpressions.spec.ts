import { describe, expect, it } from 'vitest';
import { fillExpressions } from '../regexpParseUtils.js';
import { createRegexpTokenizer } from '../regexpTokenizer.js';
import { createParserContext } from '../regexpParser.js';
import { err, ok } from '../common/fp/match.js';
import { createCharNode } from '../regexpNodeFactory.js';
import { CharType } from '../regexpNodes.js';

describe('fillExpressions', () => {
  it('should work', () => {
    const source = 'hello';
    const tokenizer = createRegexpTokenizer(source);
    const token = tokenizer.getFirstStep();
    if (!token) {
      throw new Error('Fail');
    }
    const result = fillExpressions(token, createParserContext(source, tokenizer), ({ token, nodes }) => {
      return ok({ nodes: nodes.concat(createCharNode(token.value, CharType.Simple, token)), token });
    });

    expect(result.unwrap()).toMatchSnapshot();
  });

  it('should propagate error', () => {
    const source = '12';
    const tokenizer = createRegexpTokenizer(source);
    const token = tokenizer.getFirstStep();
    if (!token) {
      throw new Error('Fail');
    }
    const result = fillExpressions(token, createParserContext(source, tokenizer), ({ token, nodes }) => {
      if (token.value === '1') {
        return ok({ nodes: nodes.concat(createCharNode(token.value, CharType.Simple, token)), token });
      }
      return err(new Error('This is 2.'));
    });

    expect(result.unwrap()).toMatchSnapshot();
  });
});
