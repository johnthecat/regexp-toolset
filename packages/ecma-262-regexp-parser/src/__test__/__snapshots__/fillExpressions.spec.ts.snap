// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fillExpressions > should propagate error 1`] = `
{
  "error": [Error: This is 2.],
  "match": false,
}
`;

exports[`fillExpressions > should work 1`] = `
{
  "match": true,
  "value": {
    "nodes": [
      {
        "charCode": 104,
        "end": 0,
        "kind": "Char",
        "start": 0,
        "type": "Simple",
        "value": "h",
      },
      {
        "charCode": 101,
        "end": 1,
        "kind": "Char",
        "start": 1,
        "type": "Simple",
        "value": "e",
      },
      {
        "charCode": 108,
        "end": 2,
        "kind": "Char",
        "start": 2,
        "type": "Simple",
        "value": "l",
      },
      {
        "charCode": 108,
        "end": 3,
        "kind": "Char",
        "start": 3,
        "type": "Simple",
        "value": "l",
      },
      {
        "charCode": 111,
        "end": 4,
        "kind": "Char",
        "start": 4,
        "type": "Simple",
        "value": "o",
      },
    ],
    "token": LinkedListNode {
      "_next": [Function],
      "_prev": [Function],
      "end": 4,
      "kind": 1,
      "start": 4,
      "value": "o",
    },
  },
}
`;
